### PR TITLE
[Merged by Bors] - feat: Bernoulli's inequality for `0 < p < 1`

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
Also substantially speed up some existing proofs in the `p < 1` case.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean [hunk 1]
```diff
@@ -147,7 +144,47 @@ theorem one_add_mul_self_le_rpow_one_add {s : ℝ} (hs : -1 ≤ s) {p : ℝ} (hp
...
+    convert strictConcaveOn_log_Ioi.secant_strict_mono this hs1 hs2 hs3 hs4 _ using 1
+    · field_simp
+    · field_simp
+    · nlinarith
+  · rw [← lt_div_iff hp1, ← div_lt_div_right hs']
+    haveI : (1 : ℝ) ∈ Ioi 0 := zero_lt_one
+    convert strictConcaveOn_log_Ioi.secant_strict_mono this hs2 hs1 hs4 hs3 _ using 1
+    · field_simp
+    · field_simp
+    · nlinarith
```
Comment (loefflerd) [inline Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean:173]: This takes 3.2 seconds to run on my machine, which is an eternity for a relatively simple statement like this one. The vast majority of that is coming from the heavyweight tactics `nlinarith` and `field_simp` which you use four times each. 

Most – possibly all – of these can be replaced with slightly uglier, but hugely quicker, longhand arguments with `rw` etc.

File: Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean [hunk 2]
```diff
@@ -147,7 +144,47 @@ theorem one_add_mul_self_le_rpow_one_add {s : ℝ} (hs : -1 ≤ s) {p : ℝ} (hp
...
+/-- **Bernoulli's inequality** for real exponents, strict version: for `0 < p < 1` and `-1 ≤ s`,
+with `s ≠ 0`, we have `(1 + s) ^ p < 1 + p * s`. -/
+theorem rpow_one_add_lt_one_add_mul_self {s : ℝ} (hs : -1 ≤ s) (hs' : s ≠ 0) {p : ℝ} (hp1 : 0 < p)
+    (hp2 : p < 1) : (1 + s) ^ p < 1 + p * s := by
+  rcases eq_or_lt_of_le hs with (rfl | hs)
+  · have : p ≠ 0 := by positivity
+    simpa [zero_rpow this]
+  have hs1 : 0 < 1 + s := by linarith
+  rcases le_or_lt (1 + p * s) 0 with hs2 | hs2
+  · nlinarith
```
Comment (loefflerd) [inline Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean:156]: Why are you writing this in such a weird roundabout way? Writing `have hs2 : 0 < 1 + p * s = by nlinarith` would be functionally equivalent and far less confusing to read.

File: Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean (hunk 1 above)
Comment (loefflerd) [reply Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean:173]: On further inspection, I see now that you are quite closely following the existing proof in the `1 ≤ p` case, which is similarly cavalier in its use of multiple heavyweight tactics. But the fact that slow code exists in the library isn't a justification for adding more bad code!

Here's a version which is structurally very similar to what you wrote, but takes something like one-sixth of the time to compile:
````
/-- **Bernoulli's inequality** for real exponents, strict version: for `0 < p < 1` and `-1 ≤ s`,
with `s ≠ 0`, we have `(1 + s) ^ p < 1 + p * s`. -/
theorem rpow_one_add_lt_one_add_mul_self {s : ℝ} (hs : -1 ≤ s) (hs' : s ≠ 0) {p : ℝ} (hp1 : 0 < p)
    (hp2 : p < 1) : (1 + s) ^ p < 1 + p * s := by
  rcases eq_or_lt_of_le hs with (rfl | hs)
  · simpa only [add_right_neg, zero_rpow hp1.ne', mul_neg, mul_one, lt_add_neg_iff_add_lt,
      zero_add]
  have hs1 : 0 < 1 + s := neg_lt_iff_pos_add'.mp hs
  have hs2 : 0 < 1 + p * s := by 
    rw [← neg_lt_iff_pos_add']
    rcases lt_or_gt_of_ne hs' with h | h
    · exact hs.trans (lt_mul_of_lt_one_left h hp2)
    · exact neg_one_lt_zero.trans (mul_pos hp1 h)
  have hs3 : 1 + s ≠ 1 := hs' ∘ add_right_eq_self.mp
  have hs4 : 1 + p * s ≠ 1 := by 
    contrapose! hs'; rwa [add_right_eq_self, mul_eq_zero, eq_false_intro hp1.ne', false_or] at hs'
  rw [rpow_def_of_pos hs1, ← exp_log hs2]
  apply exp_strictMono
  cases' lt_or_gt_of_ne hs' with hs' hs'
  · rw [← lt_div_iff hp1, ← div_lt_div_right_of_neg hs']
    convert strictConcaveOn_log_Ioi.secant_strict_mono (zero_lt_one' ℝ) hs1 hs2 hs3 hs4 _ using 1
    · rw [add_sub_cancel', div_div, log_one, sub_zero]
    · rw [add_sub_cancel', log_one, sub_zero]
    · apply add_lt_add_left (lt_mul_of_lt_one_left hs' hp2)
  · rw [← lt_div_iff hp1, ← div_lt_div_right hs']
    convert strictConcaveOn_log_Ioi.secant_strict_mono (zero_lt_one' ℝ) hs2 hs1 hs4 hs3 _ using 1
    · rw [add_sub_cancel', log_one, sub_zero]
    · rw [add_sub_cancel', div_div, log_one, sub_zero]
    · apply add_lt_add_left (mul_lt_of_lt_one_left hs' hp2)
````

I suggest you have a go at reworking the other proofs similarly. E.g. it should be possible to get `strictConcaveOn_rpow` (currently 5.8s on my machine) down to something more civilised.

Comment (Parcly-Taxel) [reply Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean]: Done.

**Comment by loefflerd** [comment]
!bench

**Comment by loefflerd** [comment]
Not bad: the file now takes about half as long as before, despite having roughly twice as many theorems! Thanks for taking on this job. I'll try to find time to proofread it for style later today, then it should be good to go.

File: Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean [hunk 3]
```diff
@@ -147,57 +138,117 @@ theorem one_add_mul_self_le_rpow_one_add {s : ℝ} (hs : -1 ≤ s) {p : ℝ} (hp
...
-      field_simp at H ⊢
-      linear_combination -H
+  · have q : 0 < y - x := by rwa [sub_pos]
+    rw [div_lt_iff q, ← div_lt_div_right hy', _root_.sub_div, div_self hy'.ne', ← div_rpow hx hy.le,
+      sub_lt_comm, ← add_sub_cancel_right (x / y) 1, add_comm, add_sub_assoc, ← div_mul_eq_mul_div,
+      mul_div_assoc, ← rpow_sub hy, sub_sub_cancel_left, rpow_neg_one, mul_assoc, ← div_eq_inv_mul,
+      sub_eq_add_neg, ← mul_neg, ← neg_div, neg_sub, _root_.sub_div, div_self hy.ne']
+    apply one_add_mul_self_lt_rpow_one_add _ _ hp
+    · rw [le_sub_iff_add_le, add_left_neg, div_nonneg_iff]; exact Or.inl ⟨hx, hy.le⟩
+    · rw [sub_ne_zero]; exact ((div_lt_one hy).mpr hxy).ne
```
Comment (loefflerd) [inline Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean:199]:
```suggestion
    · rw [le_sub_iff_add_le, add_left_neg, div_nonneg_iff]
      exact Or.inl ⟨hx, hy.le⟩
    · rw [sub_ne_zero]
      exact ((div_lt_one hy).mpr hxy).ne
```
(same again in next proof)

File: Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean [hunk 4]
```diff
@@ -147,57 +138,117 @@ theorem one_add_mul_self_le_rpow_one_add {s : ℝ} (hs : -1 ≤ s) {p : ℝ} (hp
...
+  exact (rpow_one_add_lt_one_add_mul_self hs hs' hp1 hp2).le
+
+/-- For `p : ℝ` with `1 < p`, `fun x ↦ x ^ p` is strictly convex on $[0, +∞)$.
 
 We give an elementary proof rather than using the second derivative test, since this lemma is
 needed early in the analysis library. -/
-theorem strictConvexOn_rpow {p : ℝ} (hp : 1 < p) : StrictConvexOn ℝ (Ici 0) fun x : ℝ => x ^ p := by
+theorem strictConvexOn_rpow {p : ℝ} (hp : 1 < p) : StrictConvexOn ℝ (Ici 0) fun x : ℝ ↦ x ^ p := by
   apply strictConvexOn_of_slope_strict_mono_adjacent (convex_Ici (0 : ℝ))
   rintro x y z (hx : 0 ≤ x) (hz : 0 ≤ z) hxy hyz
```
Comment (loefflerd) [inline Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean:188]:
```suggestion
  intro x y z (hx : 0 ≤ x) (hz : 0 ≤ z) hxy hyz
```
(here and below)

File: Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean [hunk 5]
```diff
@@ -147,57 +138,117 @@ theorem one_add_mul_self_le_rpow_one_add {s : ℝ} (hs : -1 ≤ s) {p : ℝ} (hp
   exact (one_add_mul_self_lt_rpow_one_add hs hs' hp).le
 #align one_add_mul_self_le_rpow_one_add one_add_mul_self_le_rpow_one_add
 
-/- For `p : ℝ` with `1 < p`, `fun x ↦ x ^ p` is strictly convex on $[0, +∞)$.
+/-- **Bernoulli's inequality** for real exponents, strict version: for `0 < p < 1` and `-1 ≤ s`,
+with `s ≠ 0`, we have `(1 + s) ^ p < 1 + p * s`. -/
+theorem rpow_one_add_lt_one_add_mul_self {s : ℝ} (hs : -1 ≤ s) (hs' : s ≠ 0) {p : ℝ} (hp1 : 0 < p)
+    (hp2 : p < 1) : (1 + s) ^ p < 1 + p * s := by
+  rcases eq_or_lt_of_le hs with (rfl | hs)
+  · rwa [add_right_neg, zero_rpow hp1.ne', mul_neg, mul_one, lt_add_neg_iff_add_lt, zero_add]
```
Comment (loefflerd) [inline Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean:146]:
```suggestion
  rcases eq_or_lt_of_le hs with rfl | hs
  · rwa [add_right_neg, zero_rpow hp1.ne', mul_neg_one, lt_add_neg_iff_add_lt, zero_add]
```
(slight golf, and formatting consistency)

File: Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean [hunk 6]
```diff
@@ -147,57 +138,117 @@ theorem one_add_mul_self_le_rpow_one_add {s : ℝ} (hs : -1 ≤ s) {p : ℝ} (hp
...
+theorem convexOn_rpow {p : ℝ} (hp : 1 ≤ p) : ConvexOn ℝ (Ici 0) fun x : ℝ ↦ x ^ p := by
   rcases eq_or_lt_of_le hp with (rfl | hp)
   · simpa using convexOn_id (convex_Ici _)
   exact (strictConvexOn_rpow hp).convexOn
 #align convex_on_rpow convexOn_rpow
 
+/-- For `p : ℝ` with `0 < p < 1`, `fun x ↦ x ^ p` is strictly concave on $[0, +∞)$.
+
+We give an elementary proof rather than using the second derivative test, since this lemma is
+needed early in the analysis library. -/
```
Comment (loefflerd) [inline Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean:219]: Is this true? I don't think this lemma's needed at all as of now (early or late). More to the point, I think proof-irrelevance applies to the reference manual: the docstring should only describe the theorem statement.

File: Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean [hunk 7]
```diff
@@ -147,57 +138,117 @@ theorem one_add_mul_self_le_rpow_one_add {s : ℝ} (hs : -1 ≤ s) {p : ℝ} (hp
...
+  rcases eq_or_lt_of_le hp2 with (rfl | hp2)
+  · simp
+  by_cases hs' : s = 0
+  · simp [hs']
+  exact (rpow_one_add_lt_one_add_mul_self hs hs' hp1 hp2).le
+
+/-- For `p : ℝ` with `1 < p`, `fun x ↦ x ^ p` is strictly convex on $[0, +∞)$.
 
 We give an elementary proof rather than using the second derivative test, since this lemma is
 needed early in the analysis library. -/
```
Comment (loefflerd) [inline Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean:185]: (see comment below)

Comment (Parcly-Taxel) [reply Mathlib/Analysis/Convex/SpecificFunctions/Basic.lean]: Done.

**Comment by loefflerd** [comment]
Umm, I just spotted something I should have realised much earlier.

There is a file `Mathlib.Analysis.Convex.SpecificFunctions.Pow`, which contains a theorem 
````
lemma strictConcaveOn_rpow {p : ℝ} (hp₀ : 0 < p) (hp₁ : p < 1) :
    StrictConcaveOn ℝ≥0 univ fun x : ℝ≥0 ↦ x ^ p := by
````
So we are just reinventing the wheel here. ☹️ I apologise for not having realised this first time around.

Perhaps the best way forward is the following. The Bernoulli inequality for p < 1 is a statement of interest in its own right, so I suggest keeping the proof you've written for that. But it might be best to replace your proof of `strictConcaveOn_rpow`, and its non-strict version, with a (hopefully much quicker) version deducing it from the existing `NNReal` version. This would have to live in `Convex.SpecificFunctions.Pow`, so I also suggest adding a comment to `Convex.SpecificFunctions.Basic` explicitly pointing out that the `p < 1` case is treated elsewhere.

**Comment by Parcly-Taxel** [comment]
> But it might be best to replace your proof of `strictConcaveOn_rpow`, and its non-strict version, with a (hopefully much quicker) version deducing it from the existing `NNReal` version. This would have to live in `Convex.SpecificFunctions.Pow`, so I also suggest adding a comment to `Convex.SpecificFunctions.Basic` explicitly pointing out that the `p < 1` case is treated elsewhere.

The proof of `Real.strictConcaveOn_rpow` already goes through the `NNReal` version:
```lean
lemma strictConcaveOn_rpow {p : ℝ} (hp₀ : 0 < p) (hp₁ : p < 1) :
    StrictConcaveOn ℝ (Set.Ici 0) fun x : ℝ ↦ x ^ p := by
  refine ⟨convex_Ici _, fun x hx y hy hxy a b ha hb hab => ?_⟩
  let x' : ℝ≥0 := ⟨x, hx⟩
  let y' : ℝ≥0 := ⟨y, hy⟩
  let a' : ℝ≥0 := ⟨a, by positivity⟩
  let b' : ℝ≥0 := ⟨b, by positivity⟩
  have hx' : (fun z => z ^ p) x = (fun z => z ^ p) x' := rfl
  have hy' : (fun z => z ^ p) y = (fun z => z ^ p) y' := rfl
  have hxy' : x' ≠ y' := Subtype.ne_of_val_ne hxy
  have hab' : a' + b' = 1 := by ext; simp [a', b', hab]
  rw [hx', hy']
  exact (NNReal.strictConcaveOn_rpow hp₀ hp₁).2 (Set.mem_univ x') (Set.mem_univ y')
    hxy' (mod_cast ha) (mod_cast hb) hab'
```

I think the better approach is to move the currently un-namespaced `(strict)?ConvexOn_rpow` _to_ `Analysis.Convex.SpecificFunctions.Pow` (where `(strict)?ConcaveOn_sqrt` also lives), namespace it and prove it from a new `NNReal version`.

**Comment by loefflerd** [comment]
> > But it might be best to replace your proof of `strictConcaveOn_rpow`, and its non-strict version, with a (hopefully much quicker) version deducing it from the existing `NNReal` version. This would have to live in `Convex.SpecificFunctions.Pow`, so I also suggest adding a comment to `Convex.SpecificFunctions.Basic` explicitly pointing out that the `p < 1` case is treated elsewhere.
> 
> The proof of `Real.strictConcaveOn_rpow` in `Analysis.Convex.SpecificFunctions.Pow` already goes through the `NNReal` version:

If this already exists in the other file, then there is even less for this PR to do. 

> I think the better approach is to move the currently un-namespaced `(strict)?ConvexOn_rpow` _to_ `Analysis.Convex.SpecificFunctions.Pow` (where `(strict)?ConcaveOn_sqrt` also lives), namespace it and prove it from a new `NNReal` version.

The reasons for not doing that are very clearly explained in the existing code. The file `Analysis.Convex.SpecificFunctions.Pow` needs more imports than `Analysis.Convex.SpecificFunctions.Basic` does. Because `strictConvexOn_rpow` is foundational for a whole load of analysis / measure theory code (anything involving L^p spaces), it is important to prove it with minimal imports; but this is not true of the concavity result for `p < 1` which has far fewer applications.

**Review by loefflerd — APPROVED** [review]
LGTM. Please update the title of this PR to match the content – when that's done I'll ping the maintainers.

**Comment by Parcly-Taxel** [comment]
> LGTM. Please update the title of this PR to match the content – when that's done I'll ping the maintainers.

It is done now?

**Comment by loefflerd** [comment]
I added a little more explanation to the header string. Code LGTM.

maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**